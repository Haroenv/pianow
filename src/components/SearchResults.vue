<template lang="pug">
  .search-results
    slot(name="header" :hits="hits")
      h2.search-results__title {{ index }} ({{ hits.length }})
    .search-results__noresults(v-if="!hits.length && noResultsMessage") {{ noResultsMessage }}
    ul.search-results__list(v-if="hits.length")
      li.search-results__item(v-for="hit in hits")
        a.search-results__link(href="#" @click.prevent="selectHit(hit)")
          slot(:hit="hit")
            div(v-html="hit._highlightResult.name.value")
    slot(name="footer" :hits="hits")
</template>

<script>
export default {
  inject: ["_helper", "_index"],
  props: {
    index: {
      type: String,
      default() {
        // use parent injected index by default
        return this._index;
      }
    },
    noResultsMessage: {
      type: String,
      default: "No results found."
    },
    options: {
      type: Object,
      default: null
    }
  },
  data() {
    return {
      localHelper: null,
      hits: []
    };
  },
  created() {
    this.localHelper = this._helper.derive(searchParams => {
      return searchParams.setIndex(this.index);
    });
    this.localHelper.on("result", content => {
      this.hits = content.hits;
    });
  },
  methods: {
    selectHit(hit) {
      // customize search results item click behavior
      this.$emit("hit:selected", hit);
    }
  }
};
</script>

<style scoped>
.search-results {
  padding: 10px;
}
.search-results__list {
  padding-left: 0;
}
.search-results__item {
  list-style: none;
}
.active {
  background-color: darkgreen;
  color: white;
  list-style: none;
}
</style>
