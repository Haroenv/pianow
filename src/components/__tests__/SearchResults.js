import Vue from "vue";
import SearchResults from "../SearchResults";

function createVm(helper, index, noResultsMessage, options) {
  return new Vue({
    components: {
      SearchResults
    },
    data() {
      return {
        index,
        noResultsMessage,
        options
      };
    },
    provide: {
      _helper: helper,
      _index: "index"
    },
    template: `<search-results :index="index" :no-results-message="noResultsMessage" :options="options"/>`
  }).$mount();
}

describe("SearchResults.vue", () => {
  const helperOn = jest.fn();
  const helper = {
    derive: jest.fn(() => ({
      on: () => helperOn()
    }))
  };

  beforeEach(() => {
    helperOn.mockClear();
    helper.derive.mockClear();
  });

  it("should derive the parent helper only once on created", () => {
    createVm(helper, "testIndex", "Custom message", {});
    expect(helper.derive).toHaveBeenCalledTimes(1);
  });

  it("should register the helper event on result", () => {
    createVm(helper, "testIndex", "Custom message", {});
    expect(helperOn).toHaveBeenCalledTimes(1);
  });

  it("should show custom no results message if there is no hits", () => {
    const vm = createVm(helper, "testIndex", "Custom message", {});
    expect(
      vm.$el.querySelector(".search-results__noresults").textContent
    ).toEqual("Custom message");
  });

  it("should show default no results message if noResultsMessage is not set", () => {
    const vm = createVm(helper, "testIndex", undefined, {});
    expect(vm.$el.querySelector(".search-results__noresults").textContent).toBe(
      "No results found."
    );
  });

  it("should not show no results message if noResultsMessage is empty", () => {
    const vm = createVm(helper, "testIndex", "", {});
    expect(vm.$el.querySelector(".search-results__noresults")).toBeNull();
  });

  it("should match the snapshot", () => {
    const vm = createVm(helper, "testIndex", undefined, {});
    expect(vm.$el).toMatchSnapshot();
  });

  describe("mergeOptions", () => {
    const options = {
      random: "test"
    };
    const vm = createVm(helper, "testIndex", "", options);
    it("should return an object with correct merged properties from data scope", () => {
      const searchParameters = {
        test: "random"
      };
      const result = vm.$children[0].mergeOptions(searchParameters);
      expect(result).toEqual(Object.assign(searchParameters, options));
    });
  });
});
