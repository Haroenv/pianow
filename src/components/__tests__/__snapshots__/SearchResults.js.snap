// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults.vue should match the snapshot 1`] = `
<div
  class="search-results"
>
  <h2
    class="search-results__title"
  >
    testIndex (0)
  </h2>
  <div
    class="search-results__noresults"
  >
    No results found.
  </div>
  <!---->
</div>
`;
